yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+  }
+
+  /**
+   * Stop the service; this triggers flushing the queue and, if not already processed,
+   * a pushing out of an application end event.
+   *
+   * This operation will block for up to `maxTimeToWaitOnShutdown` milliseconds
+   * to await the asynchronous action queue completing.
+   */
+  override protected def serviceStop {
+    // if the queue is live
+    if (!stopped.get) {
+
+      if (!appEndEventProcessed.get()) {
+        // push out an application stop event if none has been received
+        logDebug("Generating a SparkListenerApplicationEnd during service stop()")
+        val current = now()
+        innerEnqueue(new HandleSparkEvent(SparkListenerApplicationEnd(current), current))
+      }
+
+      // flush
+      logInfo(s"Shutting down: pushing out ${actionQueue.size} events")
+      innerEnqueue(FlushTimelineEvents())
+
+      // stop operation
+      postThreadActive synchronized {
+        innerEnqueue(StopQueue())
+        // now await that marker flag
+        postThreadActive.wait(maxTimeToWaitOnShutdown)
+      }
+
+      if (!actionQueue.isEmpty) {
+        // likely cause is ATS not responding.
+        // interrupt the thread, albeit at the risk of app stop events
+        // being lost. There's not much else that can be done if
+        // ATS is being unresponsive
+        logWarning(s"Did not finish flushing actionQueue before " +
+          s"stopping ATSService, eventQueueBacklog= ${actionQueue.size}")
+        eventHandlingThread.foreach(_.interrupt())
+      }
+      stopTimelineClient
+      logInfo(s"Stopped: $this")
+    }
+  }
+
+  /**
+   * Get the current entity, creating one on demand if needed, setting up its entity Id
+   * from the current appID and attempt ID values, filters from the current name and user
+   * @return the current entity
+   */
+  private def getOrCreateCurrentEntity() = {
+    curEntity.getOrElse {
+      currentEventCount = 0
+      val entity = createTimelineEntity(
+        _applicationId,
+        _attemptId,
+        sparkApplicationId,
+        sparkApplicationAttemptId,
+        appStartEventProcessed.get(),
+        appName,
+        userName)
+      curEntity = Some(entity)
+      logDebug(s"Demand creation of new entity ${describeEntity(entity)}")
+      entity
+    }
+  }
+
+  /**
+   * Perform any preflight checks.
+   * @param entity timeline entity to review.
+   */
+  private[yarn] def preflightCheck(entity: TimelineEntity): Unit = {
+    require(entity.getStartTime != null,
+      s"No start time in ${describeEntity(entity)}")
+  }
+
+  /**
+   * If there is any available entity to be sent, push to timeline server
+   */
+  private def flushEntity(): Unit = {
+    if (entityList.nonEmpty) {
+      val count = flushCount.incrementAndGet()
+      logDebug(s"flushEntity #$count: list size ${entityList.size}")
+      var client = getTimelineClient
+      val lastUpdated = now()
+      // attempt to post each entity, removing from the list all of those
```

Comment:
That's a good question; that's some code I inherited without looking at that case too closely. And right now we haven't seen/created the failures to really stress the case.

When you post an entity with a list of events, you are creating entity if needed, updating it with the new events if not. If two posts apply in a different order, the two event lists will also be out of order. Which presumably means that they are going to come in in a different order, aren't they?

Even though they come with a timestamp, so could in theory be re-ordered in the history server, that's not something you'd want to do: it would require all events to be held in memory raw and then sorted prior to replay: it doesn't scale, and isn't compatible with windowed event GET operations, which is also critical for scalability.

What should we do here? We can change that posting loop to stop on a failure, and retry later (which actually highlights a related issue: when to retry. If there are outstanding events to post, it should be attempted with some frequency (after handling every event).

I'll attempt that and see how well it works.

And yes, backpressure. I don't what the strategy should be there. Stop accepting new events exception application halt? Discard early ones? Be selective about which events to filter?